### PR TITLE
Add item recipes to config

### PR DESCRIPTION
**Pull request by Dracbjorn**
[com.chebgonaz.FriendlySkeletonWand.txt](https://github.com/jpw1991/Friendly-Skeleton-Wand/files/10369263/com.chebgonaz.FriendlySkeletonWand.txt)
### Item.cs:

- Added new InternalName class as well as a `CraftingTable` enum used to map a config string value to the actual id value in game.  This same format can potentially be used for other various settings later on.
```cs
    public class InternalName : Attribute
    {
        public readonly string internalName;
        public InternalName(string internalName) => this.internalName = internalName;
    }

    public enum CraftingTable
    {
        None,
        [InternalName("piece_workbench")] Workbench,
        [InternalName("piece_cauldron")] Cauldron,
        [InternalName("forge")] Forge,
        [InternalName("piece_artisanstation")] ArtisanTable,
        [InternalName("piece_stonecutter")] StoneCutter
    }
```
- Referenced [Azumatt's Mjolnir ItemManager](https://github.com/AzumattDev/Mjolnir/blob/master/Managers/ItemManager.cs) for help.

### SkeletonWand.cs:

- Created a new function `addMaterial(string material)` for adding materials as well as splitting the config string up.
```cs
               // Add a material to the recipe
                void addMaterial(string material)
                {
                    string[] materialSplit = material.Split(':');
                    string materialName = materialSplit[0];
                    int materialAmount = int.Parse(materialSplit[1]);
                    config.AddRequirement(new RequirementConfig(materialName, materialAmount, materialAmount*2));
                }
                
                config.CraftingStation = ((InternalName)typeof(CraftingTable).GetMember(craftingStationRequired.Value.ToString())[0].GetCustomAttributes(typeof(InternalName)).First()).internalName;

                // Material config format ex: Wood:5,Stone:1,Resin:1
                if (craftingCost.Value.Contains(','))
                {
                    string[] materialList = craftingCost.Value.Split(',');
                    foreach (string material in materialList)
                    {
                        addMaterial(material);
                    }                   
                } else
                {
                    addMaterial(craftingCost.Value);
                }                
 ```
- Added `craftingStationRequired`, `craftingStationLevel`, and `craftingCost` to config as server values. 
- I added a *2 multiplier to upgrade costs...maybe too high?  But I mean it's just wood and stone...
- Crafting cost is configured via a comma seperated list of `item:amount` pairs i.e. `wood:5,stone:1,resin:1`.

My config example:
```
## Crafting station where Skeleton Wand is available
# Setting type: CraftingTable
# Default value: Workbench
# Acceptable values: None, Workbench, Cauldron, Forge, ArtisanTable, StoneCutter
Crafting Station = Forge

## Crafting station level required to craft Skeleton Wand
# Setting type: Int32
# Default value: 1
Crafting Station Level = 1

## Materials needed to craft Skeleton Wand
# Setting type: String
# Default value: Wood:5
Crafting Costs = Wood:5,Stone:1
```

### Other

- Handled an error in `publish.ps1`

```
1>  C:\Users\Kentg\source\repos\Friendly-Skeleton-Wand\\publish.ps1 C:\Program Files (x86)\Steam\steamapps\common\Valheim 
1>  Test-Path : Cannot bind argument to parameter 'Path' because it is an empty string.
1>  At C:\Users\Kentg\source\repos\Friendly-Skeleton-Wand\publish.ps1:35 char:21
1>  +     if (!(Test-Path "$_")) {Write-Error -ErrorAction Stop -Message "$ ...
1>  +                     ~~~~
1>      + CategoryInfo          : InvalidData: (:) [Test-Path], ParameterBindingValidationException
1>      + FullyQualifiedErrorId : ParameterArgumentValidationErrorEmptyStringNotAllowed,Microsoft.PowerShell.Commands.Test 
1>     PathCommand
1>   
```

- Exposed crafting configuration to Skeleton Wand, Draugr Wand, Necro Helm and Cloak, as well as the materials to build the Spirit Pylon.
- As a result of the above changes, I fixed the error in issue #26 
- Fixed a bug referenced in issue #36 that would not allow you to set the minimum bone fragment drop amount to 0 without removing all drops.  So ranges of like 0-10 should now work.
- Exposed quality tiers (star levels) to the config for all 3 tiers. The feature was requested in issue #30 
- Exposed the necromancy skill level required for each tier to the config.  This feature wasn't requested, but I figured I might as well.

### Testing

Generate the new config.  Then mess with the crafting material requirements for all craftable items as well as the spirit pylon.  Change the workbench the item is craftable on and the workbench level requirement.  Attempt to disable them as well if you want.  It should be craftable at the Forge with my config (be default it is still the workbench).  Then ensure the crafting cost is 5 wood and 1 stone (default is 5 wood).  Then you can remove those values and test it with default values.

Change the necromancy skill level required for each tier and also mess with changing the tier quality and be sure that all adjusts accordingly.

### CODE REVIEW

**Comment by Dracbjorn** [comment]
Totally forgot there is also other stuff to craft other than the wand lol...so I will update all the other recipes to use this method before we merge and push those changes.

**Comment by Dracbjorn** [comment]
Just need to expose quality/level reqs to other minion types and this should be good to go.

File: FriendlySkeletonWand/Items/NecromancerHood.cs [hunk 1]
```diff
@@ -14,35 +14,54 @@ internal class NecromancerHood : Item
...
             base.CreateConfigs(plugin);
 
-            allowed = plugin.Config.Bind("Server config", "NecromancerHoodAllowed",
+            allowed = plugin.Config.Bind("NecromancerHood (Server Synced)", "NecromancerHoodAllowed",
                 true, new ConfigDescription("Whether crafting a Necromancer's Hood is allowed or not.", null,
                 new ConfigurationManagerAttributes { IsAdminOnly = true }));
 
-            necromancySkillBonus = plugin.Config.Bind("Server config", "NecromancerHoodSkillBonus",
-                10, new ConfigDescription("How much wearing the item should raise the Necromancy level (set to 0 to have no set effect at all).", null,
+            craftingStationRequired = plugin.Config.Bind("NecromancerHood (Server Synced)", "Necromancer Hood Crafting Station",
```
Comment (jpw1991) [inline FriendlySkeletonWand/Items/NecromancerHood.cs:29]: I think the spaces in the 2nd part should be removed. I don't think they cause any problems, but it just breaks with the convention I've seen established in other mods. So `NecromancerHoodCraftingStation` would be better even if it is a little more unreadable. It's also the way all the other configs in this mod are set up.

Granted, I have not gone and surveyed dozens of mods and how they do their configs. But the ones I've seen have all been like that.

**Review by jpw1991 — COMMENTED** [review]
Overall I'm very happy with your changes. I got a few lil nitpicks here and there. Beyond that, everything needs to be tested before I approve.

When you unmark it as draft I will do the testing and final look over.

File: FriendlySkeletonWand/Items/SkeletonWand.cs [hunk 2]
```diff
@@ -67,6 +74,18 @@ public override void CreateConfigs(BaseUnityPlugin plugin)
...
+            craftingStationRequired = plugin.Config.Bind("Server config", "Crafting Station",
+                CraftingTable.Workbench, new ConfigDescription("Crafting station where Skeleton Wand is available", null,
+                new ConfigurationManagerAttributes { IsAdminOnly = true }));
+
+            craftingStationLevel = plugin.Config.Bind("Server config", "Crafting Station Level",
+                1, new ConfigDescription("Crafting station level required to craft Skeleton Wand", null,
+                new ConfigurationManagerAttributes { IsAdminOnly = true }));
+
+            craftingCost = plugin.Config.Bind("Server config", "Crafting Costs",
+                "Wood:5", new ConfigDescription("Materials needed to craft Skeleton Wand", null,
```
[the remark this answers is not in the document]
Comment (Dracbjorn) [reply FriendlySkeletonWand/Items/SkeletonWand.cs:90]: Done for all 5 items.

```
            craftingCost = plugin.Config.Bind("SkeletonWand (Server Synced)", "Skeleton Wand Crafting Costs",
                defaultCraftingCost, new ConfigDescription("Materials needed to craft Skeleton Wand. None or Blank will use Default settings.", null,
                new ConfigurationManagerAttributes { IsAdminOnly = true }));
```

**Review by jpw1991 — APPROVED** [review]
seems to work

**Pull request merged**